File: templates/problems/06_linked_list/07_add_two_numbers/typescript/add_two_numbers.test.ts
import { describe, it, expect } from "vitest";
import { buildList, listToSlice } from "./helpers";
import { addTwoNumbers } from "./add_two_numbers";

describe("addTwoNumbers", () => {
    it("both zero", () => {
        expect(listToSlice(addTwoNumbers(buildList([0]), buildList([0])))).toEqual([0]);
    });

    it("no carry", () => {
        expect(listToSlice(addTwoNumbers(buildList([2, 4, 3]), buildList([5, 6, 4])))).toEqual([7, 0, 8]);
    });

    it("with carry", () => {
        expect(listToSlice(addTwoNumbers(buildList([9, 9, 9]), buildList([1])))).toEqual([0, 0, 0, 1]);
    });

    it("different lengths", () => {
        expect(listToSlice(addTwoNumbers(buildList([1, 8]), buildList([0])))).toEqual([1, 8]);
    });

    it("single digits", () => {
        expect(listToSlice(addTwoNumbers(buildList([5]), buildList([5])))).toEqual([0, 1]);
    });

    it("large carry chain", () => {
        expect(listToSlice(addTwoNumbers(buildList([9, 9, 9, 9]), buildList([1])))).toEqual([0, 0, 0, 0, 1]);
    });
});

File: templates/problems/06_linked_list/07_add_two_numbers/go/add_two_numbers_test.go
package add_two_numbers

import (
	"reflect"
	"testing"
)

func TestAddTwoNumbers(t *testing.T) {
	tests := []struct {
		name string
		l1   []int
		l2   []int
		want []int
	}{
		{"both zero", []int{0}, []int{0}, []int{0}},
		{"no carry", []int{2, 4, 3}, []int{5, 6, 4}, []int{7, 0, 8}},
		{"with carry", []int{9, 9, 9}, []int{1}, []int{0, 0, 0, 1}},
		{"different lengths", []int{1, 8}, []int{0}, []int{1, 8}},
		{"single digits", []int{5}, []int{5}, []int{0, 1}},
		{"large carry chain", []int{9, 9, 9, 9}, []int{1}, []int{0, 0, 0, 0, 1}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			l1 := buildList(tt.l1)
			l2 := buildList(tt.l2)
			got := listToSlice(addTwoNumbers(l1, l2))
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("addTwoNumbers() = %v, want %v", got, tt.want)
			}
		})
	}
}
